### PR TITLE
[Op] RMSNorm reference implementation

### PR DESCRIPTION
**Pull request by mitruska**
### Details:
 - RMSNorm reference implementation
 - The `Scale` input is optional
 
** The conversion of the input/output type is not included (proposed as `computation_type` - conversion logic can be handled by the Convert operations (or the reference can be extended with the conversion logic separately if agreed on the spec).
 
 Related PRs:
 - Specification Proposal: https://github.com/openvinotoolkit/openvino/pull/23569

### Tickets:
 - 136262

### CODE REVIEW

File: src/core/reference/include/openvino/reference/rms_norm.hpp [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+        std::vector<T>(reduced_elements_count, static_cast<T>(eps)).data(),
+        root_mean_square.data(),
+        reduced_shape.size());
+    sqrt(root_mean_square.data(), root_mean_square.data(), root_mean_square.size());
+    divide(in, root_mean_square.data(), out, in_shape, reduced_shape, op::AutoBroadcastType::NUMPY, false);
+
+    if (scale) {
+        multiply(out, scale, out, in_shape, scale_shape, op::AutoBroadcastType::NUMPY);
+    }
+}
```
Comment (praasz) [inline src/core/reference/include/openvino/reference/rms_norm.hpp:60]:
```suggestion
template <class T>
void rms_norm(const T* in,
              const AxisSet& axes,
              T* out,
              const Shape& in_shape,
              double eps) {
    const auto in_elem_count = shape_size(in_shape);
    std::vector<T> deg_value(in_elem_count, 2);
    power(in, deg_value.data(), out, in_elem_count);

    const auto reduced_shape = util::reduce_keep_dims(in_shape, axes);
    const auto reduced_elements_count = shape_size(reduced_shape);
    std::vector<T> root_mean_square(reduced_elements_count);
    reduce_mean(out, root_mean_square.data(), in_shape, axes);

    add(root_mean_square.data(),
        std::vector<T>(reduced_elements_count, static_cast<T>(eps)).data(),
        root_mean_square.data(),
        reduced_shape.size());
    sqrt(root_mean_square.data(), root_mean_square.data(), root_mean_square.size());
    divide(in, root_mean_square.data(), out, in_shape, reduced_shape, op::AutoBroadcastType::NUMPY, false);
}

template <class T>
void rms_norm(const T* in,
              const AxisSet& axes,
              T* out,
              const Shape& in_shape,
              double eps,
              const Shape& scale_shape,
              const T* scale) {
      rms_norm(in, axes, out, in_shape, eps);
      multiply(out, scale, out, in_shape, scale_shape, op::AutoBroadcastType::NUMPY);
}          
```

Consider add dedicated functions then `scale_shape` and `scale` data will be mandatory

**Review by praasz — CHANGES_REQUESTED** [review]

File: src/core/reference/include/openvino/reference/rms_norm.hpp (hunk 1 above)
Comment (mitruska) [reply src/core/reference/include/openvino/reference/rms_norm.hpp]: Good idea, done

File: src/plugins/template/backend/ops/rms_norm.cpp [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+
+#include "openvino/reference/rms_norm.hpp"
+
+#include "evaluate_node.hpp"
+#include "openvino/core/axis_set.hpp"
+#include "openvino/core/rank.hpp"
+#include "openvino/core/validation_util.hpp"
+#include "openvino/op/rms_norm.hpp"
+#include "openvino/op/util/axes_util.hpp"
+#include "openvino/reference/mvn.hpp"
```
[the remark this answers is not in the document]
Comment (mitruska) [reply src/plugins/template/backend/ops/rms_norm.cpp]: Right, removed

File: src/core/reference/include/openvino/reference/rms_norm.hpp [hunk 3]
```diff
@@ -19,44 +19,56 @@ namespace reference {
...
+ *
+ *  Math Formula: (x / Sqrt(ReduceMean(x^2, axes) + eps)) * scale
+ *
+ * @param in           Input pointer to data.
+ * @param axes         Axes for reduce mean calculation
+ * @param out          Output pointer to results.
+ * @param in_shape     Shape of the input Tensor
+ * @param eps          Epsilon for not dividing by zero while normalizing the value
+ * @param scale_shape  Shape of the scale Tensor (optional)
+ * @param scale        Input pointer to scale (optional).
```
Comment (praasz) [inline src/core/reference/include/openvino/reference/rms_norm.hpp:61]:
```suggestion
 * @param scale_shape  Shape of the scale Tensor.
 * @param scale        Input pointer to scale.
```

**Review by praasz — APPROVED** [review]

**Review by mmikolajcz — APPROVED** [review]

**Pull request merged**